File: src/main/java/com/baidu/algorithm/dp/_120_Triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        if triangle is None:
            return 0

        dp = [triangle[0][0]]
        for i in range(1, len(triangle)):
            dp.append(dp[-1] + triangle[i][-1])
            for j in range(len(triangle[i]) - 2, 0, -1):
                dp[j] = min(dp[j - 1], dp[j]) + triangle[i][j]
            dp[0] += triangle[i][0]

        return min(dp)

File: src/main/java/com/baidu/algorithm/dp/_120_Triangle.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dp;

import java.util.ArrayList;
import java.util.List;

/**
 * _120_Triangle
 *
 * @author xuhaoran01
 */
public class _120_Triangle {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle == null || triangle.isEmpty()) {
            return 0;
        }

        List<Integer> minPath = new ArrayList<>(triangle.get(0));
        for(int i = 1; i < triangle.size(); i++) {
            List<Integer> cur = triangle.get(i);
            minPath.add(cur.get(i) + minPath.get(i - 1));
            for(int j = minPath.size() - 2; j > 0; j--) {
                minPath.set(j, cur.get(j) + Math.min(minPath.get(j), minPath.get(j - 1)));
            }
            minPath.set(0, cur.get(0) + minPath.get(0));
        }

        int res = Integer.MAX_VALUE;
        for(int v: minPath) {
            res = Math.min(res, v);
        }

        return res;
    }

    public static void main(String[] args) {

    }
}
